### PR TITLE
Fix + add transformation example to PagingSample tests

### PR DESCRIPTION
**Pull request by dlam**

### CODE REVIEW

File: PagingSample/app/build.gradle [hunk 1]
```diff
@@ -50,25 +50,29 @@ android {
         jvmTarget = "1.8"
         freeCompilerArgs += ["-Xopt-in=kotlin.RequiresOptIn"]
     }
+
```
Comment (ChrisCraik) [inline PagingSample/app/build.gradle:53]: what are these changes for?

File: PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/CheeseViewModelTest.kt [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+            updateCallback = noopListUpdateCallback,
+            mainDispatcher = testDispatcher,
+            workerDispatcher = testDispatcher,
+        )
+        val cheeseViewModel = CheeseViewModel(CheeseDaoFake(cheeses))
+
+        // submitData allows differ to receive data from PagingData, but suspends until
+        // invalidation, so we must launch this in a separate job.
+        val job = launch {
+            println("ASD")
```
Comment (ChrisCraik) [inline PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/CheeseViewModelTest.kt:52]: remove?

File: PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/MainActivityTest.kt [hunk 3]
```diff
@@ -19,30 +19,24 @@ import android.content.Intent
...
-import org.hamcrest.MatcherAssert
+import com.google.common.truth.Truth.assertThat
 import org.junit.Test
-import org.junit.runner.RunWith
-import java.util.concurrent.TimeoutException
 
 /**
  * Simply sanity test to ensure that activity launches without any issues and shows some data.
  */
-@RunWith(AndroidJUnit4::class)
```
Comment (ChrisCraik) [inline PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/MainActivityTest.kt:32]: why is this removed?

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseDb.kt [hunk 4]
```diff
@@ -57,8 +57,7 @@ abstract class CheeseDb : RoomDatabase() {
     }
 }
 
-
-private val CHEESE_DATA = arrayListOf(
+val CHEESE_DATA = arrayListOf(
```
Comment (ChrisCraik) [inline PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseDb.kt:60]: why this change?

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseItem.kt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+package paging.android.example.com.pagingsample
+
+/**
+ * Common UI model between the [Cheese] data class and separators.
+ */
+sealed class CheeseItem(val name: String) {
```
Comment (ChrisCraik) [inline PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseItem.kt:6]: CheeseItem doesn't seem like a clear name, since it implies it's always a cheese.

Maybe CheeseListItem, to try and convey that it's an item in the cheese list, but not necessarily a cheese?

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt [hunk 6]
```diff
@@ -16,27 +16,28 @@
...
 package paging.android.example.com.pagingsample
 
-import androidx.recyclerview.widget.RecyclerView
 import android.view.LayoutInflater
 import android.view.ViewGroup
 import android.widget.TextView
+import androidx.recyclerview.widget.RecyclerView
 
 /**
  * A simple ViewHolder that can bind a Cheese item. It also accepts null items since the data may
```
Comment (ChrisCraik) [inline PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt:25]: CheeseListItem

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt [hunk 7]
```diff
@@ -16,27 +16,28 @@
...
-    var cheese : Cheese? = null
 
     /**
      * Items might be null if they are not paged in yet. PagedListAdapter will re-bind the
      * ViewHolder when Item is loaded.
      */
-    fun bindTo(cheese : Cheese?) {
-        this.cheese = cheese
-        nameView.text = cheese?.name
+    fun bindTo(item: CheeseItem?) {
```
Comment (ChrisCraik) [inline PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt:39]: what about the letter? are the rows visually distinct?

suggest doing something like:

```
if ( item is CheeseViewItem.Separator) {
    nameView.text = "${item.letter} Cheeses"
    nameView.bold = true
} else {
    nameView.text = item.name
    nameView.bold = false
}
```

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewModel.kt [hunk 8]
```diff
@@ -16,25 +16,23 @@
...
 package paging.android.example.com.pagingsample
 
-import android.app.Application
 import androidx.lifecycle.AndroidViewModel
-import androidx.paging.Pager
-import androidx.paging.PagingConfig
-import androidx.paging.PagingData
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import androidx.paging.*
```
Comment (ChrisCraik) [inline PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewModel.kt:22]: do we normally use * imports in samples?

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/MainActivity.kt [hunk 9]
```diff
@@ -61,9 +61,17 @@ class MainActivity : AppCompatActivity() {
...
             // enable the items to swipe to the left or right
-            override fun getMovementFlags(recyclerView: RecyclerView,
-                                          viewHolder: RecyclerView.ViewHolder): Int =
-                makeMovementFlags(0, ItemTouchHelper.LEFT or ItemTouchHelper.RIGHT)
+            override fun getMovementFlags(
+                recyclerView: RecyclerView,
+                viewHolder: RecyclerView.ViewHolder
+            ): Int {
+                val cheeseViewHolder = viewHolder as CheeseViewHolder
+                return if (cheeseViewHolder.cheese != null) {
```
Comment (ChrisCraik) [inline PagingSample/app/src/main/java/paging/android/example/com/pagingsample/MainActivity.kt:69]: simplify these 2 lines as:

`return if (viewHolder is CheeseViewHolder.Item)`

File: PagingSample/app/build.gradle (hunk 1 above)
Comment (dlam) [reply PagingSample/app/build.gradle:53]: This is a competing kotlin / coroutine versions thing where the same license is getting added multiple times :|

File: PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/CheeseViewModelTest.kt (hunk 2 above)
Comment (dlam) [reply PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/CheeseViewModelTest.kt:52]: Done

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseDb.kt (hunk 4 above)
Comment (dlam) [reply PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseDb.kt:60]: Reverted.

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseItem.kt (hunk 5 above)
Comment (dlam) [reply PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseItem.kt:6]: Done

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt (hunk 7 above)
Comment (dlam) [reply PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt:39]: Hmm this is probably more instructional, good idea.

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewModel.kt (hunk 8 above)
Comment (dlam) [reply PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewModel.kt:22]: Hmmm good point - I think the default code styles just does this, but I just updated my settings locally to fix this.

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/MainActivity.kt (hunk 9 above)
Comment (dlam) [reply PagingSample/app/src/main/java/paging/android/example/com/pagingsample/MainActivity.kt:69]: Actually we only have one viewholder for both :|

File: PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/MainActivityTest.kt (hunk 3 above)
Comment (dlam) [reply PagingSample/app/src/androidTest/java/paging/android/example/com/pagingsample/MainActivityTest.kt:32]: Oops - done, was getting a test not found issue earlier so was testing this temporarily :|

File: PagingSample/app/build.gradle (hunk 1 above)
Comment (dlam) [reply PagingSample/app/build.gradle]: Yeah I'm really not sure, but it seems to fail consistent in circle ci without it.

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt (hunk 6 above)
Comment (ChrisCraik) [reply PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt:26]: Still need to fix this comment, either "can bind a Cheese or Separator item" or "can bind a CheeseListItem"

**Review by ChrisCraik — APPROVED** [review]

File: PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt (hunk 6 above)
Comment (dlam) [reply PagingSample/app/src/main/java/paging/android/example/com/pagingsample/CheeseViewHolder.kt]: Done
